modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -17,94 +17,156 @@
  */
 public class IncomingSnapshotCopier extends SnapshotCopier {
-    /** Messages factory. */
+    private static final IgniteLogger LOG = Loggers.forClass(IncomingSnapshotCopier.class);
+
     private static final TableMessagesFactory MSG_FACTORY = new TableMessagesFactory();
 
-    /** {@link SnapshotStorage} instance for the partition. */
-    private final PartitionSnapshotStorage snapshotStorage;
+    private static final long NETWORK_TIMEOUT = 10_000;
 
-    /** Snapshot URI. */
-    private final SnapshotUri snapshotUri;
+    private final PartitionSnapshotStorage partitionSnapshotStorage;
 
-    /** Rebalance thread-pool, used to write data into a storage. */
-    //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-    // Use external pool.
-    private final ExecutorService threadPool = Executors.newSingleThreadExecutor();
+    private final SnapshotUri snapshotUri;
 
     /**
      * Snapshot meta read from the leader.
      *
      * @see SnapshotMetaRequest
      */
-    private SnapshotMeta snapshotMeta;
+    @Nullable
+    private volatile SnapshotMeta snapshotMeta;
+
+    private volatile boolean canceled;
+
+    @Nullable
+    private volatile CompletableFuture<?> future;
 
     /**
      * Constructor.
      *
-     * @param snapshotStorage Snapshot storage.
+     * @param partitionSnapshotStorage Snapshot storage.
      * @param snapshotUri Snapshot URI.
      */
-    public IncomingSnapshotCopier(PartitionSnapshotStorage snapshotStorage, SnapshotUri snapshotUri) {
-        this.snapshotStorage = snapshotStorage;
+    public IncomingSnapshotCopier(PartitionSnapshotStorage partitionSnapshotStorage, SnapshotUri snapshotUri) {
+        this.partitionSnapshotStorage = partitionSnapshotStorage;
         this.snapshotUri = snapshotUri;
     }
 
     @Override
-    public void cancel() {
-        //TODO https://issues.apache.org/jira/browse/IGNITE-17262
-        // Implement.
+    public void start() {
+        Executor executor = partitionSnapshotStorage.getIncomingSnapshotsExecutor();
+
+        LOG.info("Copier is started for the partition: " + partId());
+
+        future = prepareMvPartitionStorageForRebalance(executor)
+                .thenCompose(unused -> prepareTxStatePartitionStorageForRebalance(executor))
+                .thenCompose(unused -> {
+                    ClusterNode snapshotSender = getSnapshotSender(snapshotUri.nodeName);
+
+                    if (snapshotSender == null) {
+                        LOG.error("Snapshot sender not found [partId={}, nodeName={}]", partId(), snapshotUri.nodeName);
+
+                        if (!isOk()) {
+                            setError(RaftError.UNKNOWN, "Sender node was not found or it is offline");
+                        }
+
+                        return completedFuture(null);
+                    }
+
+                    return loadSnapshotMeta(snapshotSender)
+                            .thenCompose(unused1 -> loadSnapshotMvData(snapshotSender, executor))
```

Comment:
But this would be a useless argument, why do this? Maybe I didn't understand you.

Reply:
Both options looks weird, so it's up to you to decide.
Maybe you should use `thenRun`, it has no unused argument.
Another thing I noticed is that you always pass executor as argument into methods like "loadSnapshotMvData". Can't it be a field?

Reply:
I'll leave it as is.
`thenRun` does not return the future, but I need it so that the cancel works correctly with an interrupt.
I can even not use the field, it seems to me that it's a little clearer that we will execute the task in another thread.

Reply:
You're right, I forgot about it

Reply:
> I can even not use the field

Why?